Live Chat

nick-funk opened the pull request:
<!--

Thank you for submitting a pull request! Please note that by contributing to
Coral, you agree to our Code of Conduct: http://code-of-conduct.voxmedia.com/

Before submitting your Pull Request (or PR), please verify that:

* [ ] Your code is up-to-date with the base branch
* [ ] You've successfully run `npm run test` locally

Refer to CONTRIBUTING.MD for more details.

  https://github.com/coralproject/talk/blob/main/CONTRIBUTING.md

-->

## What does this PR do?

<!--

In this section, you should be describing what other Github issues or tickets
that this PR is designed to addressed.

Any related Github issue should be linked by adding its URL to this section.

-->

## What changes to the GraphQL/Database Schema does this PR introduce?

<!--

In this section, you should describe any changes to be made to the GraphQL
schema file (located https://github.com/coralproject/talk/blob/main/src/core/server/graph/schema/schema.graphql) or any
database model (located as types in the https://github.com/coralproject/talk/blob/main/src/core/server/models directory).

If no changes were added to the GraphQL/Database Schema as a part of this PR,
simply write "None".

-->

## How do I test this PR?

<!--

In this section, you should be describing any manual testing that can be used to
verify features introduced or bugs fixed in this PR.

 -->

nick-funk committed:
Create initial live tab pane

wyattjoh committed:
feat: added Story.comments(flatten)

nick-funk committed:
Get before comments loading with pagination

nick-funk committed:
Get after comments loading with pagination

nick-funk committed:
Basic scroll to load more behaviour

alinhle commented on src/core/client/stream/tabs/Live/AfterComments.tsx:
```diff
@@ -0,0 +1,127 @@
+        includeBefore: true,
+        includeAfter: true,
+        storyID: story.id,
+        flattenReplies: true,
```
Some of these are not used?

alinhle commented on src/core/client/stream/tabs/Live/LiveStream.tsx:
```diff
@@ -0,0 +1,249 @@
+        await loadMore();
+
+        if (beforeComments.length > 0) {
+          const id = `comment-${beforeComments[0].__id}`;
```
Request `id` explicitly and use that.

alinhle commented on src/core/client/stream/tabs/Live/LiveStream.tsx:
```diff
@@ -0,0 +1,249 @@
+    const end = endRef.current as any;
+    end.scrollIntoView();
+
+    setTimeout(() => {
```
clear timeout

nick-funk committed:
Use "after" as connection on AfterComments container

alinhle commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -531,6 +531,11 @@ enum FEATURE_FLAG {
+  """
+  Enables the chat UI for displaying a new Chat story mode.
+  """
+  CHAT
```
ENABLE_CHAT?

alinhle committed:
fix: cursor type

wyattjoh committed:
feat: added inclusive argument to Story.comments

nick-funk committed:
Move shared items to tabs/shared

nick-funk committed:
Load id in live comment stream/after comments

nick-funk committed:
Create preliminary LivePostCommentFormContainer

Is mostly a modified duplicate of the PostCommentFormContainer
but with R&R and Q&A ripped out.

Will create a custom LiveCreateCommentMutation next.

nick-funk committed:
Fix cursor plumbing to the after comments

nick-funk committed:
Create preliminary LiveCreateCommentMutation

nick-funk committed:
Hook up Live comment entered subscription for chat

nick-funk committed:
Fix incorrect ReactonConfig import

nick-funk committed:
Keep stream marker refs as any type

It appears that setting them to HTMLDivElement breaks
the scrolling behaviour? Weird...

nick-funk committed:
Hook up preliminary tailing behaviour

nick-funk committed:
Consolidate before and after comments into self-ingested composition

alinhle committed:
fix: withFragmentContainer types now supports plurals

alinhle committed:
fix: use hasNextPage instead of relay.hasMore() and fix subscription condition.

nick-funk committed:
When not tailing, append after comments to paginate

When user scrolls up, then comes back down to the after
comments, we will paginate and auto load more of the
after comments that are available.

Also, while the user is scrolled up, we avoid having the
performance issues of having endless comments loading in.

nick-funk committed:
Convert local storage tailing to a local relay cached field

nick-funk committed:
Show parent on live chat reply comments

nick-funk committed:
Use double value cursor state to store view state

nick-funk committed:
Scroll to before/after when cursor is set

nick-funk committed:
Create a preliminary reply modal

nick-funk committed:
Preliminarily style the reply modal elements

nick-funk committed:
Switch to using div w/ overlay for reply modal

nick-funk committed:
Fix replies not inserting into stream

nick-funk committed:
Create empty LiveCommentRepliesQuery

nick-funk committed:
Set some of the Live Comment styles

nick-funk committed:
Fix horizontal scroll bug

nick-funk committed:
Create scroll to new reply button

nick-funk committed:
Remove drafts on reply comment form

nick-funk committed:
Style before/after into new comments divider

nick-funk committed:
Style live reply comments

alinhle committed:
fix: use ie11 compatible scrolling

nick-funk committed:
Style the reply to form

nick-funk committed:
Hook up preliminary conversation button

nick-funk committed:
Create preliminary replies pagination container

nick-funk committed:
Convert live stream to use edges instead of node arrays

nick-funk committed:
Create more efficient InView component for live comments

nick-funk committed:
Set cursors on live inserted commente edges

nick-funk committed:
Remove unnecessary end ref

nick-funk committed:
Preliminarily hook up comments to conversation view

nick-funk committed:
Scroll to comment when replying to it

nick-funk committed:
Set cursor when inserting new reply edge

nick-funk committed:
Prevent infinite scrolling in live chat stream

nick-funk committed:
Hook up subscription for live chat replies view

alinhle commented on src/core/client/stream/tabs/Live/LiveChatContainer.tsx:
```diff
@@ -0,0 +1,500 @@
+          </Button>
+        </div>
+      )}
+      {replyVisible && focusedComment && storyID && viewer && (
```
```suggestion
      {replyVisible && focusedComment && viewer && (
```

nick-funk committed:
Trying to improve mobile safari scrolling

nick-funk committed:
Style reply view comments same as live stream

nick-funk committed:
Query replyCount so replies trigger conversation buttons

nick-funk committed:
Insert reply comments into reply conversation view

nick-funk committed:
Handle when user not signed in

nick-funk committed:
Use right timeout types in LiveChatContainer

nick-funk committed:
Update snapshots with live chat tab

nick-funk committed:
Fix alphabetization on import

nick-funk committed:
order imports

nick-funk committed:
Remove storyID check as it's not necessary

nick-funk committed:
Fix auto-load-more scrolling for iOS/Safari

nick-funk committed:
Jump to reply using cursor instead of scroll

wyattjoh committed:
feat: Added `Comment.replies(inclusive: Boolean)`

nick-funk committed:
Use before/after querying for live chat replies

nick-funk committed:
Hook up loadMore for before/after in live chat replies

nick-funk committed:
Add more ways to get to conversation

nick-funk committed:
Make sure jump to comment always works

- Slightly longer time out wait before scroll
- Scroll target bottom of comment, not mid-top

nick-funk committed:
Use onConversationParent not onConversation to open parent

nick-funk committed:
Fix ellipses on reply-parent body text

alinhle committed:
wip: chat rte

alinhle committed:
fix: rte css

alinhle committed:
wip: disable autoHideToolbar

alinhle committed:
feat: auto hide toolbar for live chat

nick-funk committed:
Align action buttons under comment body

nick-funk committed:
Always show conversation when replying

nick-funk committed:
Allow jumping to submitted comment in live chat

nick-funk committed:
Remove scroll to parent when direct replying

nick-funk committed:
Refactor comment actions into separate container

nick-funk committed:
Update snaps for RTE changes

nick-funk committed:
Fix reply button not targeting right level of convo

nick-funk committed:
Show rec and report buttons in conversation view

nick-funk committed:
Hook up ignore container to live chat usernames

nick-funk committed:
Add moderate buttons for live chat

nick-funk committed:
Show moderate button on comment actions

nick-funk committed:
Disable commenting actions if user banned/suspended/warned

nick-funk committed:
Add avatar support for live chat

nick-funk committed:
Create jump to reply button in conversation view

nick-funk committed:
Add close button to new comment/reply toasts

nick-funk committed:
Emit preliminary events for live chat

- Load more before (scrolling up and loading)
- When user opens a conversation thread

nick-funk committed:
Show tags and badges on live comment body

nick-funk committed:
Disable giphy for live chat

We don't have full support or designs for this yet

nick-funk committed:
Swap out Report for Moderate when staff in live chat

nick-funk committed:
Make live chat a story mode

nick-funk committed:
Show callout when there are no comments

nick-funk committed:
Add sign-in controls to live chat stream

nick-funk committed:
import ordering fix

nick-funk committed:
Handle logged out state for conversation views

nick-funk committed:
Add missing icons to live comment actions

nick-funk committed:
Update styles for main live comment parent ref's

alinhle commented on src/core/client/stream/tabs/Live/LiveChatContainer.tsx:
```diff
@@ -0,0 +1,703 @@
+      window.clearTimeout(scrollToCommentTimeout.current);
+    }
+
+    window.setTimeout(scrollToComment, 300);
```
I think you forgot to set the timeout ref:
`scrollToCommentTimeout.current = window.clearTimeout(scrollToCommentTimeout.current);`

nick-funk committed:
Update styles for username/badges/tags/timestamp

nick-funk committed:
Make conversation close button darker

alinhle committed:
chore: refactor live chat post form

alinhle commented:
Noting this here so we don't forget:
- When story is closed and comments are empty, we would still encourage user to write comments even when it's not possible.
![image](https://user-images.githubusercontent.com/14221600/110827746-a9c69080-8296-11eb-960c-cdf8dd3c95ea.png)

nick-funk committed:
Reload page if user has been away for a while

alinhle commented on src/core/client/stream/tabs/Live/LiveTabQuery.tsx:
```diff
@@ -67,6 +68,14 @@ const LiveTabQuery: FunctionComponent = () => {
+      window.location.reload();
+    }
+  }, []);
+  useHeartbeat(reloadOnIdle, heartbeatIntervalMs);
```
This looks a bit hardcore 👀

nick-funk committed:
Emit more events for live chat

- How conversation mode is opened
- When they jump to comment/reply

alinhle committed:
Move errors above chat, add focus style, and improve toolbar behavior

alinhle committed:
Fix some live chat types and add missing fragment

alinhle commented on src/core/client/stream/tabs/Live/LiveCreateCommentMutation.tsx:
```diff
@@ -0,0 +1,353 @@
+    .getRootField("createComment")!
+    .getLinkedRecord("edge")!;
+
+  commentEdge.setValue(new Date().toISOString(), "cursor");
```
`commentEdge.setValue(new Date().toISOString(), "cursor");`

alinhle committed:
Fix snapshots

nick-funk committed:
Create distinct events for 4 ways to open conversation view

- on reply to comment
- on reply to parent
- on show conversation
- on show parent conversation

nick-funk committed:
Add warning around heartbeat reload logic

nick-funk committed:
Rename handler callbacks for clarity

nick-funk committed:
Use createdAt for setting mutation cursor values

nick-funk committed:
Set missing timeout ref

nick-funk committed:
Handle no comments, story closed state in live chat

nick-funk committed:
Encapsulate heartbeat check into useOnResumeActive hook

nick-funk committed:
Fix font size and icons for live comment actions

alinhle committed:
Manage live chat layout and improve RTE resize behavior

nick-funk committed:
Make live chat jump to comment buttons more rounded

nick-funk committed:
Make tailing a little easier to trigger

nick-funk committed:
Up the inactive threshold to 3 seconds

nick-funk committed:
Up the onResumeActive threshold to 8000 ms

We don't want this refreshing randomly on slower
machines. Was possibly an issue during recent tests.

alinhle committed:
feat: use react-virtuoso

alinhle committed:
fix: subscription now pause when not tailing and hasMore

alinhle committed:
fix: when not tailing, and a new comment comes in, scroll down load correctly

alinhle committed:
fix: improved loading behavior and improved subscription stability

alinhle committed:
chore: refactor

alinhle committed:
fix: set start index

alinhle committed:
chore: throw error on unexpected index

alinhle committed:
fix: forgot to remove commented line

alinhle committed:
chore: remove unused css

alinhle committed:
fix: jump to newly posted comment

nick-funk committed:
Emit events when we start or stop tailing

alinhle committed:
fix: ignored user support in virtuoso

nick-funk committed:
Show jump to live when we have new comments and are not tailing

alinhle committed:
feat: integrate virtuoso into conversation view

alinhle committed:
fix: use global intersection provider

alinhle committed:
fix: hide non functional reply button in convo view for now

nick-funk committed:
Hide the conversation jump to reply if we see target comment

alinhle committed:
fix: improve jump to reply behavior

nick-funk committed:
Track when user posts a comment when not tailing

nick-funk committed:
Handle when before/after comments are null during reload

nick-funk committed:
Create separate jump to new and jump to live button actions

nick-funk committed:
fix: prevent blank stream when jumping to new

nick-funk committed:
Maintain live chat position on set cursor events

nick-funk committed:
Only delete connection if we can find it

nick-funk committed:
Categorize live chat events as "liveChat"

alinhle committed:
Simplify visibility of new message prompts

alinhle committed:
Improve jump to reply behavior

alinhle committed:
Report invalid connection issue and throw during dev and test

nick-funk committed:
Move liveChat categorisation to the rudder specific logging

Also breaks the open conversation events with 4 type variants
into separate events so that analytics tools like Google
Analytics can properly distinguish between these events by name.

nick-funk committed:
Add reacted classnames to live chat comments

nick-funk committed:
Disable "My comments" from Profile tab when using live chat

nick-funk committed:
Append live chat rudder events with "coral:"

nick-funk committed:
Title comments tab as "Live" when in live chat mode

nick-funk committed:
Create editing flow for main live chat stream

nick-funk committed:
Hook up editing for conversation view

nick-funk committed:
Add proper styles to edit mode title

nick-funk committed:
Improve emphasis of currently editing comment

nick-funk committed:
Only show edit if we have a valid onEdit callback

nick-funk committed:
Fix padding for edit targeted reply comments

nick-funk committed:
Show edit time remaining on edit form

nick-funk committed:
Hide Profile > My Comments when in live chat for users

Was previously dependent on feature flags which were
not available at startup to regular users.

nick-funk committed:
Do not assume we have connections in live chat subscriptions

alinhle committed:
fix: scrolling inside conversation view

nick-funk committed:
Hide preferences > notifications when in live chat mode

nick-funk committed:
Preliminary hook up of comment edit subscription events

nick-funk committed:
Add resolver for CommentEditedPayload

nick-funk committed:
Mark edited live chat comments as edited

nick-funk committed:
Forgot to sort the imports, fixing it now

alinhle committed:
fix: remove margin

alinhle committed:
fix: save cursor to local storage

alinhle committed:
chore: refactor CommentForm classnames, add todos

alinhle committed:
fix: undesired overflow in conversation view

alinhle committed:
fix: maybe this helps with gc?

nick-funk committed:
Make create/reply rate limit configurable via env variable

nick-funk committed:
Force garbage collection after setting live chat cursor

alinhle committed:
Add deleteConnection helper

nick-funk committed:
Hook up new deleteConnection logic from @cvle

nick-funk committed:
Bring back the timeout for conversation view

alinhle committed:
quick fix: conversation view failing when jumping to cursor

nick-funk committed:
Rename LiveCommentReply to LiveConversation

nick-funk committed:
When user clicks outside of conversation, close conversation

alinhle committed:
fix: stop rendering the pagination containers before clearing the connection data

nick-funk committed:
Replace dangerouslySetInnerHTML with HTMLContent

nick-funk committed:
Swap out use of timer in onResumeActive

Use the handleVisibilityChange instead of a timer
checking how long since last script execution.

alinhle committed:
chore: another refactor for CommentForm

alinhle committed:
fix: add some rte translations

nick-funk committed:
Revert "Swap out use of timer in onResumeActive"

This reverts commit dfb695c2351a72ea2981eeb0ac00003574cbafa9.

nick-funk committed:
Improve "New comments" and "Jump to live" visibility logic

Makes the buttons for jumping to new comments and jumping to
live comments less glitchy when toggling their visibility
states.

nick-funk committed:
Show warnings, suspensions, and bans in live chat

nick-funk committed:
Show live viewer count in live chat if feature flag enabled

To use this, we need to enable the VIEWER_COUNT feature
flag.

nick-funk committed:
Set showReport so Report button visually activates

This was all hooked up except for one line, which
was keeping the report button from properly
activating visually when we opened the report flow.

nick-funk committed:
Hook up jump to live button for conversation view

nick-funk committed:
Show reactions in live chat when signed out

nick-funk committed:
Disable bullet lists in live chat RTE

Adds optional override for the basic RTE features
that allows us to enable/disable the following RTE
features:

- bold
- italic
- block quote
- bullet point lists

nick-funk committed:
Hide the RTE toolbar when user isn't signed in

nick-funk committed:
Refactor live chat jump buttons into re-usable component

nick-funk committed:
Remove completed TODO's from LiveChatContainer

alinhle committed:
fix: improve JumpToX Quirks

nick-funk committed:
Store cursor state in relay

- currentCursor depicts the current "NEW" dividing cursor
we see in our live chat until we reload or jump around the
stream
- latestCursor stores the latest cursor we have actively
read to in the stream. This allows us to always access both
the current new marker we're showing as well as how far the
viewer has traversed in the stream past that point.

The latestCursor is persisted to both the local relay but
also the localStorage, so when the stream is reloaded, that
will become the new currentCursor.

Similarly, when the user jumps around in the stream via a
refresh or jump to live, the currentCursor is set along with
the pagination state set in the LiveTabQuery.

alinhle committed:
Implement behavior around banned or rejected edit and refactor

nick-funk commented:
Closing this PR because we have created an epic branch for live chat here:
https://github.com/coralproject/talk/pull/3504
